arm7/lib/src/SND_channel.c

```diff
@@ -0,0 +1,192 @@
+    reg_SOUNDoffPNT(off) = (u16)loopStart;
+    reg_SOUNDoffLEN(off) = (u32)loopLength;
+    reg_SOUNDoffSAD(off) = (u32)data;
+}
+
+void SND_SetupChannelPsg(int chnIdx, int duty, int volume, int volumeDiv, int timer, int pan)
+{
+    int off = chnIdx * 0x10;
+
+    sOrgPan[chnIdx] = (u8)pan;
+    if (sMasterPan >= 0)
+        pan = sMasterPan;
+
+    sOrgVolume[chnIdx] = (u8)volume;
+    if (sSurroundDecay > 0 && (1 << chnIdx) & 0xFFF5)
+    {
+        volume = CalcSurroundDecay(volume, pan);
+    }
+
+    reg_SOUNDoffCNT(off) =
+        (u32)(0x60000000 | (duty << 24) | (pan << 16) | (volumeDiv << 8) | volume);
+    reg_SOUNDoffTMR(off) = (u16)(0x10000 - timer);
+}
+
+void SND_SetupChannelNoise(int chnIdx, int volume, int volumeDiv, int timer, int pan)
+{
+    int off = chnIdx * 0x10;
+
+    sOrgPan[chnIdx] = (u8)pan;
+    if (sMasterPan >= 0)
+        pan = sMasterPan;
+
+    sOrgVolume[chnIdx] = (u8)volume;
+    if (sSurroundDecay > 0 && (1 << chnIdx) & 0xFFF5)
+    {
+        volume = CalcSurroundDecay(volume, pan);
+    }
+
+    reg_SOUNDoffCNT(off) = (u32)(0x60000000 | (pan << 16) | (volumeDiv << 8) | volume);
+    reg_SOUNDoffTMR(off) = (u16)(0x10000 - timer);
+}
+
+void SND_StopChannel(int idx, int hold)
+{
+    vu32 *reg = &reg_SOUNDxCNT(idx);
+
+    u32 v = *reg;
+
+    // disable channel
+    v &= ~0x80000000;
+
+    // set hold flag
+    if (hold & 1)
+        v |= 0x8000;
+
+    *reg = v;
+}
+
+void SND_SetChannelVolume(int chnIdx, int vol, int volDiv)
+{
+    sOrgVolume[chnIdx] = (u8)vol;
+
+    if (sSurroundDecay > 0 && (1 << chnIdx) & 0xFFF5)
+    {
+        int pan = reg_SOUNDxCNT_PAN(chnIdx);
+        vol = CalcSurroundDecay(vol, pan);
+    }
+
+    reg_SOUNDxCNT_VOLS(chnIdx) = (u16)((volDiv << 8) | vol);
+}
+
+void SND_SetChannelTimer(int chnIdx, int timer)
+{
+    reg_SOUNDxTMR(chnIdx) = (u16)(0x10000 - timer);
+}
+
+void SND_SetChannelPan(int chnIdx, int pan)
+{
+    sOrgPan[chnIdx] = (u8)pan;
+
+    if (sMasterPan >= 0)
+    {
+        pan = sMasterPan;
+    }
+
+    reg_SOUNDxCNT_PAN(chnIdx) = (u8)pan;
+
+    if (sSurroundDecay > 0 && (1 << chnIdx) & 0xFFF5)
+    {
+        reg_SOUNDxCNT_VOL(chnIdx) = (u8)CalcSurroundDecay(sOrgVolume[chnIdx], pan);
+    }
+}
+
+BOOL SND_IsChannelActive(int chnIdx)
+{
+    return (reg_SOUNDxCNT_STAT(chnIdx) & 0x80) != 0;
+}
+
+void SND_SetMasterPan(int pan)
+{
+    sMasterPan = pan;
+
+    if (pan >= 0)
+    {
+        for (int i = 0; i < SND_CHANNEL_COUNT; i++)
+        {
+            reg_SOUNDxCNT_PAN(i) = (u8)pan;
+        }
+    }
+    else
+    {
+        for (int i = 0; i < SND_CHANNEL_COUNT; i++)
+        {
+            reg_SOUNDxCNT_PAN(i) = sOrgPan[i];
+        }
+    }
+}
+
+u32 SND_GetChannelControl(int chnIdx)
+{
+    return reg_SOUNDxCNT(chnIdx);
+}
+
+void SNDi_SetSurroundDecay(int decay)
+{
+    sSurroundDecay = decay;
+
+    for (int i = 0; i < SND_CHANNEL_COUNT; i++)
+    {
+        // do not process channel 1+3 (capture playback channels)
+        if ((1 << i) & 0xFFF5)
```

Comment:
I'm seeing a lot of 0xFFF5, I'm not sure what this means, again, constant?

Reply:
FFF5 is a mask where bit 1 and bit 3 are cleared (see comment)